### PR TITLE
asserts,store: complete support for optional primary key headers for assertions

### PR DESCRIPTION
**Pull request by pedronis**
Teach Find and helpers dealing with primary keys to consider default values for optional primary key headers.

When interacting with the store use reduced/normalized primary keys omitting default values to help with backward compatibility.

### CODE REVIEW

File: asserts/asserts.go [hunk 1]
```diff
@@ -247,16 +277,25 @@ func SuggestFormat(assertType *AssertionType, headers map[string]interface{}, bo
...
 // primaryKey values and the assertion type, it errors if primaryKey
-// has the wrong length.
+// does not cover all the non-optional primary key headers or provides
+// too many values.
 func HeadersFromPrimaryKey(assertType *AssertionType, primaryKey []string) (headers map[string]string, err error) {
-	if len(primaryKey) != len(assertType.PrimaryKey) {
+	n := len(assertType.PrimaryKey)
+	nopt := len(assertType.OptionalPrimaryKeyDefaults)
+	ninp := len(primaryKey)
+	if ninp > n || ninp < (n-nopt) {
```
Comment (mardy) [inline asserts/asserts.go:286]: Can't this be
```suggestion
	if !assertType.AcceptablePrimaryKey(primaryKey) {
```
?

File: asserts/asserts.go [hunk 2]
```diff
@@ -287,23 +326,54 @@ func HeadersFromSequenceKey(assertType *AssertionType, sequenceKey []string) (he
...
 }
 
+// ReducePrimaryKey produces a primary key prefix by omitting any
+// suffix of optional primary key headers default values.
+// Too short or long primary keys are returned as is.
+func ReducePrimaryKey(assertType *AssertionType, primaryKey []string) []string {
+	n := len(assertType.PrimaryKey)
+	nopt := len(assertType.OptionalPrimaryKeyDefaults)
+	ninp := len(primaryKey)
+	if ninp > n || ninp < (n-nopt) {
```
Comment (mardy) [inline asserts/asserts.go:359]: same here?

**Review by mardy — APPROVED** [review]
LGTM, but CI is unhappy!

File: asserts/asserts.go [hunk 3]
```diff
@@ -287,23 +326,54 @@ func HeadersFromSequenceKey(assertType *AssertionType, sequenceKey []string) (he
...
 	for i, k := range keys {
 		keyVal := headers[k]
 		if keyVal == "" {
-			return nil, fmt.Errorf("must provide primary key: %v", k)
+			defl := defaults[k]
+			if defl != "" {
+				keyVal = defl
+			} else {
+				return nil, fmt.Errorf("must provide primary key: %v", k)
+			}
```
Comment (stolowski) [inline asserts/asserts.go:345]: Nitpick^2, how about something like:
```
if keyVal == "" {
    keyVal = defaults[k]
    if keyVal == "" {
      return nil, fmt.Errorf("must provide primary key: %v", k)
   }
}
```

**Review by stolowski — APPROVED** [review]
Looks fine, one tiny nitpick, also Alberto's suggestions make sense.

**Pull request merged**